### ISSUE TITLE
Dired-like interaction

### ISSUE THREAD

**Issue by now** [magit/magit#2384]
I would really like to interact with Magit the way I interact with Dired or VC, where the things to operate n are selected before the operation is executed.  For example, I’d like to mark the files or patches to commit, then hit x to commit.

I haven’t looked at Magit’s source, so I’d first just like to know if this kind of interaction would even be possible with Magit’s current “plumbing”.  If so, I’d at least consider working on this myself.

Thanks!

**Comment by Silex** [comment]
Is that really different than staging your diffs with `s` and then pressing `c c` when ready to commit?

**Comment by tarsius** [comment]
No, this would not be possible because Magit does not provide a mechanism for marking sections, at least not in a way that persists when the point and/or mark change.

You can mark neighboring sections of the same type by using the region though. If you mark some file (or hunk) sections this way and then stage them, then they won't be marked by prefixing them with a `*` as they would in Dired. Instead they now appear in the section "Staged changes" instead of in "Unstaged changes" -- which could be interpreted as a kind of "mark". Likewise think of staging (<kbd>s</kbd>) as "marking for commit".

**Comment by now** [comment]
I realize that there’s no support now; my ineptly stated question was whether adding such support would be possible or, rather, feasible, considering how Magit is currently implemented.

Would it be feasible to add “section marking” support?

Using Git’s idea of staging is, sadly, one level too low, as I would like to use this kind of setup to avoid extraneous calls to Git when working on Windows.

**Comment by tarsius** [comment]
Please clarify what you mean by "one level too low"?

**Comment by now** [comment]
Sorry, I just mean that, yes, Git’s staging area is sort of like what I’m after, but as it interacting with it requires running Git, the main point of what I’m after is lost.

One of the benefits of the Dired-like interaction would be that one would be able to minimize the amount of calls to Git.

**Comment by Silex** [comment]
> One of the benefits of the Dired-like interaction would be that one would be able to minimize the amount of calls to Git.

While I agree it's a benefit, I don't see at all what it would serve... also right now staging is much better than what exists in dired because you can stage _hunks_ and not the whole file.

**Comment by npostavs** [comment]
> (I mean, staging is very fast already)

You have the good fortune of not using Windows :)

I've also wanted something like this for deleting remote branches. I may be misremembering, but didn't magit once upon have something like this, where `.` did the marking?

> staging is better than what exists in dired because you can stage hunks and not the whole file.

I don't think "Dired-like" should be understood to imply working with files. It's just the marking UI that we're talking about.

**Comment by Silex** [comment]
>> staging is better than what exists in dired because you can stage hunks and not the whole file.

> I don't think "Dired-like" should be understood to imply working with files. It's just the marking UI that we're talking about.

So that means hunks are "marked" with a special color? no more separation of staged/unstaged parts?

I can see a dired-like interface for stuffs where you'd mark commits in the log and operate on them (for interactive rebases?), or maybe remotes like you said... but for staging/unstaging I just don't see it :)

**Comment by npostavs** [comment]
> no more separation of staged/unstaged parts?

It wouldn't replace staging. It's a way to batch several stage/push/whatever commands together.

**Comment by tarsius** [comment]
> I've also wanted something like this for deleting remote branches. I may be misremembering, but didn't magit once upon have something like this, where `.` did the marking?

That is correct, but it was only possible to mark a single commit. It wasn't possible to mark multiple commits or anything that isn't a commit. And that feature only existed for the benefit of a single command, "show differences between the commit at point and the marked commit". But while it provided almost no benefit (IMO that command was rather useless in the first place) it made necessary refactorings harder. So I nuked it.

>> (I mean, staging is very fast already)

> You have the good fortune of not using Windows :)

I don't think Windows being slow at starting subprocesses is the issue here. That may *contribute* to the bad performance, but I would suspect that the known issues (a) washing diffs is slow and (b) we constantly and needlessly recreate the whole buffer, even when just parts of it need refreshing, are the real issues.

I am going to tackle these issues in `2.5` and `3.1`.

> Git’s staging area is sort of like what I’m after, but as it interacting with it requires running Git, the main point of what I’m after is lost.

I think that is backward. You want the best possible staging experience. "Not using Git" is not the goal but what you suspect would solve the performance issues. I am always in favor of fixing the real issues instead of adding kludges that work around the issue, which is what you are proposing here.

So I am not going to implement "section marking", at least not now and definitely not for this purpose. The ability to mark sections like this might be of value, but not when the goal is to implement a crippled alternative to Git's staging area.

Currently I also would not merge a contributed "section marking" implementation, regardless of it's intended use-case(s). It is extremely important that we improve performance and that requires some heavy refactoring - adding such a feature would complicate the refactoring.

So I have added "Maybe implement 'section marking'" to the [Roadmap](https://github.com/magit/magit/issues/2180) for v3.1, and am closing this issue (which doesn't mean you cannot keep discussing this here). (But I am on a kind-of-break, so I might be slow at responding.)
